### PR TITLE
fix(consul): support to fetch only health endpoint

### PR DESCRIPTION
**Pull request by Fabriceli**
### Description
Thank you for your previous work @dyrnq

Create TWO threads to watch the catalog services and health services:

```mermaid
flowchart TD
		subgraph Thread 1: watch catalog
		A[Start] --> catalog{Catalog any change?}
    catalog -->|Yes| cindex[return catalog index]
    catalog ---->|No| cblock[block query until catalog any change]
    cblock --> E[End]
    cindex --> E
    end
    
    subgraph Thread 2: watch health
    A2[Start] --> health2{health any change?}
    health2 -->|Yes| hindex[return health index]
    health2 ---->|No| hblock[block query until health any change]
    hblock --> E2[End]
    hindex --> E2
    end
    
    subgraph Wait any thread
    A3[Start] --> spawn[Spawn TWO threads]
    spawn --> wait[wait any Thread]
    wait --> catalog2[Query all services]
    catalog2 --> H[Query health nodes]
    H --> update[Update all services]
    update --> e[End]
    end
```

Fixes # (issue)
Reference: #8856, https://github.com/apache/apisix/issues/8928

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by Fabriceli** [comment]
@tokers @spacewander @tzssangglass Please take a look at this PR, thanks

**Comment by Fabriceli** [comment]
@liuxiran Could you return CI? some other test error, which is not I add

**Comment by monkeyDluffy6017** [comment]
@Fabriceli The ci has been reruned

**Comment by Fabriceli** [comment]
@monkeyDluffy6017 Please take a look at this PR, thanks

**Comment by Fabriceli** [comment]
@monkeyDluffy6017 The CI is not very stable,could you re-run the CI?

**Comment by Fabriceli** [comment]
@monkeyDluffy6017 Please take a look at this PR, thanks

**Comment by Fabriceli** [comment]
the Unit test is not very stable:
```sh
#   Failed test 'ERROR: client socket timed out - t/config-center-yaml/route-upstream.t TEST 5: upstream domain
# '
#   at /home/runner/work/apisix/apisix/test-nginx/lib/Test/Nginx/Socket.pm line 2220.

#   Failed test 't/config-center-yaml/route-upstream.t TEST 5: upstream domain - status code ok'
#   at /home/runner/work/apisix/apisix/test-nginx/lib/Test/Nginx/Socket.pm line 937.
#          got: ''
#     expected: '200'
```

another fail:
```sh
#   Failed test 'ERROR: client socket timed out - t/node/merge-route.t TEST 12: hit route
# '
#   at /home/runner/work/apisix/apisix/test-nginx/lib/Test/Nginx/Socket.pm line 2220.

#   Failed test 't/node/merge-route.t TEST 12: hit route - status code ok'
#   at /home/runner/work/apisix/apisix/test-nginx/lib/Test/Nginx/Socket.pm line 950.
#          got: ''
#     expected: '200'

#   Failed test 't/node/merge-route.t TEST 12: hit route - response_body - like (repeated req 0, req 0)'
#   at /home/runner/work/apisix/apisix/test-nginx/lib/Test/Nginx/Socket.pm line 1671.
#                   ''
#     doesn't match '(?^:"Host": "httpbin.org")'
[15:39:53] t/node/merge-route.t .... 
Failed 3/68 subtests 
[15:40:02] t/xrpc/redis.t .......... ok     8996 ms ( 0.01 usr  0.00 sys +  2.17 cusr  0.80 csys =  2.98 CPU)
[15:40:02]
```

**Comment by Fabriceli** [comment]
@monkeyDluffy6017 Encore ERROR:
```sh
#   Failed test 'ERROR: client socket timed out - t/plugin/authz-casdoor.t TEST 6: test code handling
# '
#   at /apisix/test-nginx/lib/Test/Nginx/Socket.pm line [222](https://github.com/apache/apisix/actions/runs/4732449515/jobs/8426677125?pr=9204#step:16:223)0.

#   Failed test 't/plugin/authz-casdoor.t TEST 6: test code handling - status code ok'
#   at /apisix/test-nginx/lib/Test/Nginx/Socket.pm line 950.
#          got: ''
#     expected: '200'

#   Failed test 't/plugin/authz-casdoor.t TEST 6: test code handling - response_body - response is expected (repeated req 0, req 0)'
#   at /apisix/test-nginx/lib/Test/Nginx/Socket.pm line 1674.
#          got: ''
#     expected: 'done
# '
[16:06:17] t/plugin/authz-casdoor.t ..
```
Fail 2:
```sh
Error: mERROR] [ERROR] Some problems were encountered while processing the POMs:
Warning:  'build.plugins.plugin.version' for org.apache.maven.plugins:maven-compiler-plugin is missing. @ line 87, column 25
Error:  Non-resolvable import POM: Could not transfer artifact org.springframework.boot:spring-boot-dependencies:pom:2.1.5.RELEASE from/to central (https://repo.maven.apache.org/maven2): Transfer failed for https://repo.maven.apache.org/maven2/org/springframework/boot/spring-boot-dependencies/2.1.5.RELEASE/spring-boot-dependencies-2.1.5.RELEASE.pom @ line 37, column 25
Error:  Non-resolvable import POM: Could not transfer artifact org.apache.dubbo:dubbo-dependencies-bom:pom:2.7.21 from/to central (https://repo.maven.apache.org/maven2): Transfer failed for https://repo.maven.apache.org/maven2/org/apache/dubbo/dubbo-dependencies-bom/2.7.21/dubbo-dependencies-bom-2.7.21.pom @ line 44, column 25
 @ 
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]   
[ERROR]   The project org.apache.dubbo.backend:dubbo-backend:1.0.0-SNAPSHOT (/home/runner/work/apisix/apisix/t/lib/dubbo-backend/pom.xml) has 2 errors
[ERROR]     Non-resolvable import POM: Could not transfer artifact org.springframework.boot:spring-boot-dependencies:pom:2.1.5.RELEASE from/to central (https://repo.maven.apache.org/maven2): Transfer failed for https://repo.maven.apache.org/maven2/org/springframework/boot/spring-boot-dependencies/2.1.5.RELEASE/spring-boot-dependencies-2.1.5.RELEASE.pom @ line 37, column 25: Unknown host repo.maven.apache.org: Name or service not known -> [Help 2]
[ERROR]     Non-resolvable import POM: Could not transfer artifact org.apache.dubbo:dubbo-dependencies-bom:pom:2.7.21 from/to central (https://repo.maven.apache.org/maven2): Transfer failed for https://repo.maven.apache.org/maven2/org/apache/dubbo/dubbo-dependencies-bom/2.7.21/dubbo-dependencies-bom-2.7.21.pom @ line 44, column 25: Unknown host repo.maven.apache.org -> [Help 2]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException
[ERROR] [Help 2] http://cwiki.apache.org/confluence/display/MAVEN/UnresolvableModelException
Error: Process completed with exit code 1.
```

**Comment by Fabriceli** [comment]
@monkeyDluffy6017  could you return the CI, I have no any change in the file `t/plugin/authz-casdoor.t` but it failed:
```sh
#   Failed test 't/plugin/authz-casdoor.t TEST 6: test code handling - response_body - response is expected (repeated req 0, req 0)'
#   at /apisix/test-nginx/lib/Test/Nginx/Socket.pm line 1674.
#          got: ''
#     expected: 'done
# '
```
thanks

**Comment by Fabriceli** [comment]
@monkeyDluffy6017 all CI run successfully, please take a look at this PR, thanks

File: apisix/discovery/consul/init.lua [hunk 1]
```diff
@@ -175,81 +188,309 @@ local function get_retry_delay(retry_delay)
...
+        log.warn("watch catalog has no change, retry call consul after ", random_delay, " seconds")
+        core_sleep(random_delay)
+        goto RETRY
+    end
+    return watch_type_catalog, watch_result.headers['X-Consul-Index']
+end
+
+local function watch_health(consul_server)
+    local opts
+    if consul_server.keepalive then
```
[the remark this answers is not in the document]
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua]: done

File: apisix/discovery/consul/init.lua [hunk 2]
```diff
@@ -175,81 +188,309 @@ local function get_retry_delay(retry_delay)
...
+            read_timeout = consul_server.read_timeout,
+        }
+    end
+    local client = resty_consul:new(opts)
+
+    ::RETRY::
+    local watch_result, watch_err = client:get(consul_server.consul_watch_health_url)
+    local watch_error_info = (watch_err ~= nil and watch_err)
+            or ((watch_result ~= nil and watch_result.status ~= 200)
+            and watch_result.status)
```
[the remark this answers is not in the document]
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua:250]: It just check whether has error in the result, i think it's readable.

**Comment by Fabriceli** [comment]
ping @monkeyDluffy6017 @soulbird

**Comment by monkeyDluffy6017** [comment]
Hi @Fabriceli,  thank you for all your hard work!
I'm working on this pr, it will take me some time to figure it out.

File: apisix/discovery/consul/init.lua [hunk 3]
```diff
@@ -161,12 +175,11 @@ local function show_dump_file()
     if not data then
         return 503, err
     end
-
```
Comment (monkeyDluffy6017) [inline apisix/discovery/consul/init.lua:164]: Why do you remove this blank line?

Comment (Fabriceli) [reply apisix/discovery/consul/init.lua:164]: just re-fomat the code

File: apisix/discovery/consul/init.lua [hunk 4]
```diff
@@ -175,81 +188,293 @@ local function get_retry_delay(retry_delay)
...
+        local random_delay = math_random(default_random_seed)
+        log.error("failed to spawn thread watch catalog: ", spawn_catalog_err,
+            ", retry connecting consul after ", random_delay, " seconds")
+        core_sleep(random_delay)
+
+        check_keepalive(consul_server, retry_delay)
+        return
+    end
+
+    local health_thread, err = thread_spawn(watch_health, consul_server)
```
Comment (monkeyDluffy6017) [inline apisix/discovery/consul/init.lua:343]: Could you add some comments to explain why you need to watch these 2 uri at the same time?

File: apisix/discovery/consul/init.lua [hunk 5]
```diff
@@ -175,81 +188,293 @@ local function get_retry_delay(retry_delay)
...
+                index = consul_server.health_index,
+            }
+        end
+    end
+
+    return opts
+end
+
+local function watch_catalog(consul_server)
+
```
Comment (monkeyDluffy6017) [inline apisix/discovery/consul/init.lua:216]: remove useless blank

File: apisix/discovery/consul/init.lua (hunk 4 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua:343]: The goal is to get all change: after read all consul APIs, Get all change = health change + catalog change
so we need to watch two url. If any change, we need to update the all service table[service_name: ip/port_info]

Reference：
- https://github.com/apache/apisix/issues/8856
- https://discuss.hashicorp.com/t/help-get-any-change-of-the-service-via-consul-api/51957

File: apisix/discovery/consul/init.lua (hunk 5 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua]: done

File: apisix/discovery/consul/init.lua [hunk 6]
```diff
@@ -175,81 +188,293 @@ local function get_retry_delay(retry_delay)
...
+            and consul_server.health_index == tonumber(watch_result.headers['X-Consul-Index']) then
+        local random_delay = math_random(default_random_seed)
+        log.info("watch health has no change, re-watch consul after ", random_delay, " seconds")
+        core_sleep(random_delay)
+        goto RETRY
+    end
+    return watch_type_health, watch_result.headers['X-Consul-Index']
+end
+
+local function check_keepalive(consul_server, retry_delay)
```
Comment (monkeyDluffy6017) [inline apisix/discovery/consul/init.lua:267]: The functions needs to be separated by two blank lines, please check other similar places

File: apisix/discovery/consul/init.lua [hunk 7]
```diff
@@ -175,81 +188,291 @@ local function get_retry_delay(retry_delay)
...
+            and watch_result.status)
+    if watch_error_info then
+        log.error("connect consul: ", consul_server.consul_server_url,
+            " by sub url: ", consul_server.consul_watch_catalog_url,
+            ", got watch result: ", json_delay_encode(watch_result, true),
+            ", with error: ", watch_error_info)
+
+        return watch_type_catalog, default_catalog_error_index
+    end
+    if consul_server.catalog_index > 0
```
Comment (monkeyDluffy6017) [inline apisix/discovery/consul/init.lua:231]: Could you add a blank between the two blocks?

File: apisix/discovery/consul/init.lua [hunk 8]
```diff
@@ -175,81 +188,291 @@ local function get_retry_delay(retry_delay)
...
+            and consul_server.catalog_index == tonumber(watch_result.headers['X-Consul-Index']) then
+        local random_delay = math_random(default_random_seed)
+        log.info("watch catalog has no change, re-watch consul after ", random_delay, " seconds")
+        core_sleep(random_delay)
+        goto RETRY
+    end
+    return watch_type_catalog, watch_result.headers['X-Consul-Index']
```
Comment (monkeyDluffy6017) [inline apisix/discovery/consul/init.lua:238]: Add a blank

File: apisix/discovery/consul/init.lua (hunk 3 above)
Comment (monkeyDluffy6017) [reply apisix/discovery/consul/init.lua:164]: I think we can keep this, like other place in you file

File: apisix/discovery/consul/init.lua [hunk 9]
```diff
@@ -175,81 +188,291 @@ local function get_retry_delay(retry_delay)
...
+            and watch_result.status)
+    if watch_error_info then
+        log.error("connect consul: ", consul_server.consul_server_url,
+            " by sub url: ", consul_server.consul_watch_health_url,
+            ", got watch result: ", json_delay_encode(watch_result, true),
+            ", with error: ", watch_error_info)
+
+        return watch_type_health, default_health_error_index
+    end
+    if consul_server.health_index > 0
```
Comment (monkeyDluffy6017) [inline apisix/discovery/consul/init.lua:257]: Add a blank

File: apisix/discovery/consul/init.lua [hunk 10]
```diff
@@ -175,81 +188,291 @@ local function get_retry_delay(retry_delay)
...
+            and consul_server.health_index == tonumber(watch_result.headers['X-Consul-Index']) then
+        local random_delay = math_random(default_random_seed)
+        log.info("watch health has no change, re-watch consul after ", random_delay, " seconds")
+        core_sleep(random_delay)
+        goto RETRY
+    end
+    return watch_type_health, watch_result.headers['X-Consul-Index']
```
Comment (monkeyDluffy6017) [inline apisix/discovery/consul/init.lua:264]: Add a blank

File: apisix/discovery/consul/init.lua [hunk 11]
```diff
@@ -175,81 +188,291 @@ local function get_retry_delay(retry_delay)
...
+
+        check_keepalive(consul_server, retry_delay)
+        return
+    end
+
+    local thread_wait_ok, watch_type, index = thread_wait(catalog_thread, health_thread)
+    thread_kill(catalog_thread)
+    thread_kill(health_thread)
+    if not thread_wait_ok then
+        log.error("failed to wait thread: ", watch_type)
```
Comment (monkeyDluffy6017) [inline apisix/discovery/consul/init.lua:357]: duplicate error log

File: apisix/discovery/consul/init.lua [hunk 12]
```diff
@@ -175,81 +188,291 @@ local function get_retry_delay(retry_delay)
...
+        core_sleep(random_delay)
+
+        check_keepalive(consul_server, retry_delay)
+        return
+    end
+
+    -- double check index has changed
+    if not watch_result_is_valid(tonumber(watch_type),
+            tonumber(index), consul_server.catalog_index, consul_server.health_index) then
+
```
Comment (monkeyDluffy6017) [inline apisix/discovery/consul/init.lua:369]: Remove useless blank

File: apisix/discovery/consul/init.lua [hunk 13]
```diff
@@ -175,81 +188,291 @@ local function get_retry_delay(retry_delay)
...
 
     log.info("connect consul: ", consul_server.consul_server_url,
-        ", watch_result status: ", watch_result.status,
-        ", watch_result.headers.index: ", watch_result.headers['X-Consul-Index'],
+        ", catalog_result status: ", catalog_res.status,
+        ", catalog_result.headers.index: ", catalog_res.headers['X-Consul-Index'],
         ", consul_server.index: ", consul_server.index,
         ", consul_server: ", json_delay_encode(consul_server, true))
 
     -- if current index different last index then update service
```
Comment (monkeyDluffy6017) [inline apisix/discovery/consul/init.lua:444]: if the current index is different from the last index, then update the service

File: apisix/discovery/consul/init.lua [hunk 14]
```diff
@@ -175,81 +188,291 @@ local function get_retry_delay(retry_delay)
...
-        for service_name, _ in pairs(watch_result.body) do
+    if (consul_server.catalog_index ~= tonumber(catalog_res.headers['X-Consul-Index']))
+            or (consul_server.health_index ~= tonumber(health_res.headers['X-Consul-Index'])) then
+        local up_services = core.table.new(0, #catalog_res.body)
+        for service_name, _ in pairs(catalog_res.body) do
             -- check if the service_name is 'skip service'
             if skip_service_map[service_name] then
                 goto CONTINUE
             end
             -- get node from service
```
Comment (monkeyDluffy6017) [inline apisix/discovery/consul/init.lua:453]: Add a blank

File: apisix/discovery/consul/init.lua [hunk 15]
```diff
@@ -175,81 +188,291 @@ local function get_retry_delay(retry_delay)
...
                 local nodes = up_services[service_name]
-                for  _, node in ipairs(result.body) do
-                    local svc_address, svc_port = node.ServiceAddress, node.ServicePort
-                    if not svc_address then
-                        svc_address = node.Address
+                for _, node in ipairs(result.body) do
+                    if not node.Service then
+                        goto CONTINUE
                     end
+                    local svc_address, svc_port = node.Service.Address, node.Service.Port
```
Comment (monkeyDluffy6017) [inline apisix/discovery/consul/init.lua:475]: Add a blank

File: apisix/discovery/consul/init.lua [hunk 16]
```diff
@@ -175,81 +188,291 @@ local function get_retry_delay(retry_delay)
...
+            ", got health result: ", json_delay_encode(health_res, true),
+            ", with error: ", health_error_info)
 
         retry_delay = get_retry_delay(retry_delay)
-        log.warn("retry connecting consul after ", retry_delay, " seconds")
+        log.warn("get all svcs got err, retry connecting consul after ", retry_delay, " seconds")
         core_sleep(retry_delay)
 
-        goto ERR
+        goto ERROR
```
Comment (monkeyDluffy6017) [inline apisix/discovery/consul/init.lua:435]: Why not write it like this?
```
        check_keepalive(consul_server, retry_delay)
        return
```

File: t/discovery/consul.t [hunk 17]
```diff
@@ -582,3 +582,78 @@ GET /thc
 [{"ip":"127.0.0.1","port":30513,"status":"healthy"},{"ip":"127.0.0.1","port":30514,"status":"healthy"}]
 [{"ip":"127.0.0.1","port":30513,"status":"healthy"},{"ip":"127.0.0.1","port":30514,"status":"healthy"}]
 --- ignore_error_log
+
+
+
+=== TEST 13: test consul catalog service change
```
Comment (monkeyDluffy6017) [inline t/discovery/consul.t:588]: Could you add a test case to prove that `fetch only health endpoint` now?

File: apisix/discovery/consul/init.lua (hunk 16 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua]: done

File: apisix/discovery/consul/init.lua (hunk 15 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua:490]: done

File: apisix/discovery/consul/init.lua (hunk 14 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua:467]: done

File: apisix/discovery/consul/init.lua (hunk 13 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua]: done

File: apisix/discovery/consul/init.lua (hunk 6 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua:276]: done

File: apisix/discovery/consul/init.lua (hunk 7 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua:235]: done

File: apisix/discovery/consul/init.lua (hunk 8 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua:243]: done

File: apisix/discovery/consul/init.lua (hunk 9 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua:264]: done

File: apisix/discovery/consul/init.lua (hunk 10 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua:272]: done

File: apisix/discovery/consul/init.lua (hunk 11 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua]: done

File: apisix/discovery/consul/init.lua [hunk 18]
```diff
@@ -176,80 +190,304 @@ local function get_retry_delay(retry_delay)
...
+            }
+        else
+            opts.default_args = {
+                wait = consul_server.wait_timeout, --blocked wait!=0; unblocked by wait=0
+                index = consul_server.health_index,
+            }
+        end
+    end
+
+    return opts
```
Comment (leslie-tsang) [inline apisix/discovery/consul/init.lua:214]:
```suggestion
    if not consul_server.keepalive then
        return opts
    end
    
    if is_catalog then
        opts.default_args = {
            wait = consul_server.wait_timeout, --blocked wait!=0; unblocked by wait=0
            index = consul_server.catalog_index,
        }
    else
        opts.default_args = {
            wait = consul_server.wait_timeout, --blocked wait!=0; unblocked by wait=0
            index = consul_server.health_index,
        }
    end

    return opts
```
Would be better ?

File: apisix/discovery/consul/init.lua [hunk 19]
```diff
@@ -176,80 +190,304 @@ local function get_retry_delay(retry_delay)
...
+
+
+local function watch_catalog(consul_server)
+    local client = resty_consul:new(get_opts(consul_server, true))
+
+    ::RETRY::
+    local watch_result, watch_err = client:get(consul_server.consul_watch_catalog_url)
+    local watch_error_info = (watch_err ~= nil and watch_err)
+            or ((watch_result ~= nil and watch_result.status ~= 200)
+            and watch_result.status)
```
Comment (leslie-tsang) [inline apisix/discovery/consul/init.lua:225]:
```suggestion
    local watch_error_info = (watch_err ~= nil and watch_err)
                             or ((watch_result ~= nil and watch_result.status ~= 200)
                             and watch_result.status)
```
Plz follow the [CODE_STYLE](https://github.com/apache/apisix/blob/master/CODE_STYLE.md#maximum-length-per-line)

File: apisix/discovery/consul/init.lua [hunk 20]
```diff
@@ -176,80 +190,304 @@ local function get_retry_delay(retry_delay)
...
+
+    ::RETRY::
+    local watch_result, watch_err = client:get(consul_server.consul_watch_health_url)
+    local watch_error_info = (watch_err ~= nil and watch_err)
+            or ((watch_result ~= nil and watch_result.status ~= 200)
+            and watch_result.status)
+    if watch_error_info then
+        log.error("connect consul: ", consul_server.consul_server_url,
+            " by sub url: ", consul_server.consul_watch_health_url,
+            ", got watch result: ", json_delay_encode(watch_result, true),
```
Comment (leslie-tsang) [inline apisix/discovery/consul/init.lua:258]:
```suggestion
            ", got watch result: ", json_delay_encode(watch_result, true),
```
No need for `json_delay_encode` as the log level is `error` by default ?

File: apisix/discovery/consul/init.lua [hunk 21]
```diff
@@ -44,9 +52,15 @@ local events_list
 local consul_services
 
 local default_skip_services = {"consul"}
+local default_random_seed = 5
```
Comment (leslie-tsang) [inline apisix/discovery/consul/init.lua:55]: Would a new variable name be better?
It confuses me as a `seed` as it don't do the seed stuff, would be better to like `range` or something else ?

File: apisix/discovery/consul/init.lua [hunk 22]
```diff
@@ -176,80 +190,304 @@ local function get_retry_delay(retry_delay)
...
+        consul_server.health_index = h_index
+    end
+end
+
+
+local function is_not_empty(value)
+    if value == nil or value == null
+            or (type(value) == "table" and not next(value))
+            or (type(value) == "string" and value == "") then
+        return false
```
Comment (leslie-tsang) [inline apisix/discovery/consul/init.lua:312]:
```suggestion
    if value == nil or value == null
       or (type(value) == "table" and not next(value))
       or (type(value) == "string" and value == "") 
    then
        return false
```
Make it more easier to read. :)

**Review by leslie-tsang — COMMENTED** [review]
Thanks for your contribution. :)

File: apisix/discovery/consul/init.lua (hunk 3 above)
Comment (monkeyDluffy6017) [reply apisix/discovery/consul/init.lua:164]: @Fabriceli

File: apisix/discovery/consul/init.lua (hunk 20 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua]: you are right, done

File: apisix/discovery/consul/init.lua (hunk 19 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua]: done

File: apisix/discovery/consul/init.lua (hunk 18 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua]: done

File: apisix/discovery/consul/init.lua (hunk 3 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua]: done

File: apisix/discovery/consul/init.lua (hunk 22 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua]: done

File: apisix/discovery/consul/init.lua (hunk 21 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua]: done

**Review by monkeyDluffy6017 — APPROVED** [review]

File: apisix/discovery/consul/init.lua (hunk 20 above)
Comment (leslie-tsang) [reply apisix/discovery/consul/init.lua]:
```suggestion
            ", got watch result: ", json_encode(watch_result, true),
```
Sry for the mistake

**Pull request merged**